### PR TITLE
node 0.10, more checks for keylen

### PR DESCRIPTION
**Pull request by fanatid**
Another attempt (summary of #22 #23 #25 #26)

### CODE REVIEW

**Comment by dcousens** [comment]
I haven't got time to merge this at the moment.  Once-over looks good.
ping @calvinmetcalf

**Comment by calvinmetcalf** [comment]
this does much more then simply update the keylen checks, can you add some comments in the pull to make it easeir to review the changes ?

**Comment by fanatid** [comment]
- `index.js` is more clean now, when node v0.10 supports will be finished (01.2017?) it will be easy update this package
- `browser.js` additionally exports `_checkParameters` and `_createHmac` (it's done for node v0.10), by default `_createHmac` uses pure js hmac (but `node-shim.js` update `_createHmac` for using `crypto.createHmac`)

@calvinmetcalf if you have any questions, please ask!

File: browser.js [hunk 1]
```diff
@@ -34,20 +14,33 @@ function checkParameters (iterations, keylen) {
     throw new TypeError('Key length not a number')
   }
 
-  if (keylen < 0 || keylen > MAX_ALLOC) {
+  if (keylen < 0 || isNaN(keylen) || keylen > MAX_ALLOC) {
```
Comment (calvinmetcalf) [inline browser.js]: probably want to do `keylen !== keylen` as older browsers aren't going to have isNaN

File: browser.js [hunk 2]
```diff
@@ -1,27 +1,7 @@
...
-    throw new Error('No callback provided to pbkdf2')
-  }
-
-  setTimeout(function () {
-    var result = pbkdf2Sync(password, salt, iterations, keylen, digest)
-
-    callback(null, result)
-  })
-}
+exports._createHmac = require('create-hmac/browser')
```
Comment (calvinmetcalf) [inline browser.js]: why is this exported? can't it be imported again?  Also why does this need to be the /browser, those are mainly intended for internal testing purposes and could totally change in the future (but at the moment require('create-hmac') and  require('create-hmac/browser') resolve to be the same in the browser).

File: package.json [hunk 3]
```diff
@@ -19,8 +19,8 @@
     "coverage": "istanbul cover _mocha -- -t 20000 test/index.js",
     "coveralls": "npm run coverage && coveralls < coverage/lcov.info",
     "standard": "standard",
-    "test": "mocha --reporter list -t 20000 test/index.js",
-    "bundle-test": "browserify test/index.js > test/bundle.js"
+    "test": "mocha --reporter spec -t 20000 test/index.js",
+    "test:bundle": "browserify test/index.js >test/browser.js"
```
Comment (calvinmetcalf) [inline package.json]: why is this changed?

File: browser.js (hunk 1 above)
Comment (fanatid) [inline browser.js]: in `lodash` `isNaN`: `value != +value`
https://github.com/lodash/lodash/blob/092f90d2fca959fa75d14cae8a2377a58bcb122d/lodash.js#L10534
or better will use `keylen !== keylen` ?

File: package.json (hunk 3 above)
Comment (fanatid) [inline package.json]: I think it's more logical, it's test related task, but honestly I don't care about names. Maybe I should put `karma`?

File: browser.js (hunk 2 above)
Comment (fanatid) [inline browser.js]: `create-hmac/browser` require `create-hash` in the same way, the main purpose all of this -- get pure js `pbkdf2`, `create-hmac`, `create-hash`

File: browser.js (hunk 1 above)
Comment (calvinmetcalf) [inline browser.js]: lodash is making sure it's a number, but we already have a check to do that

File: package.json (hunk 3 above)
Comment (calvinmetcalf) [inline package.json]: well not karma because we aren't doing karma there, we don't tend to use colons to separate out related tasks so that might be a bit confusing why don't we change that back and then open a separate issue to change it

File: README.md [hunk 4]
```diff
@@ -1,15 +1,17 @@
 # pbkdf2
 
-[![build status](https://secure.travis-ci.org/crypto-browserify/pbkdf2.png)](http://travis-ci.org/crypto-browserify/pbkdf2)
-[![Coverage Status](https://img.shields.io/coveralls/crypto-browserify/pbkdf2.svg)](https://coveralls.io/r/crypto-browserify/pbkdf2)
-[![Version](http://img.shields.io/npm/v/pbkdf2.svg)](https://www.npmjs.org/package/pbkdf2)
+[![NPM Package](https://img.shields.io/npm/v/pbkdf2.svg?style=flat-square)](https://www.npmjs.org/package/pbkdf2)
```
Comment (calvinmetcalf) [inline README.md]: can we have these in their own issue?

Comment (calvinmetcalf) [inline README.md]: by issue I mean pull

File: async-shim.js [hunk 5]
```diff
@@ -1,19 +0,0 @@
-var compat = require('./browser')
```
Comment (calvinmetcalf) [inline async-shim.js]: I'm not against the renaming of this file, but it does make the diff harder to see, can we have the name changed back and then we can do a separate pull to rename the file?

File: README.md (hunk 4 above)
Comment (fanatid) [inline README.md]: done. #28

File: browser.js (hunk 1 above)
Comment (fanatid) [inline browser.js]: done

File: package.json (hunk 3 above)
Comment (fanatid) [inline package.json]: put old code

File: async-shim.js (hunk 5 above)
Comment (fanatid) [inline async-shim.js]: renamed

File: browser.js (hunk 2 above)
Comment (calvinmetcalf) [inline browser.js]: yes but why do you need pure pbkdf2 create-hmac and create-hash?

Comment (fanatid) [inline browser.js]: for example [cryptocoinjs/scryptsy](https://github.com/cryptocoinjs/scryptsy)
but if you think that this isn't necessary, we should remove `create-hash/browser` in `create-hmac`

Comment (calvinmetcalf) [inline browser.js]: so the idea behind the /browser is so we can test the browser version in node. 

Since as far as I can tell, the actual outcome of this code is the same result as the must less complex situation of having a `var createHmac = require('create-hmac')` here I don't feel like we are gaining anything.

File: browser.js [hunk 6]
```diff
@@ -34,20 +14,33 @@ function checkParameters (iterations, keylen) {
...
-function pbkdf2Sync (password, salt, iterations, keylen, digest) {
-  checkParameters(iterations, keylen)
+exports.pbkdf2 = function (password, salt, iterations, keylen, digest, callback) {
+  if (typeof digest === 'function') {
+    callback = digest
+    digest = undefined
+  }
 
-  digest = digest || 'sha1'
+  exports._checkParameters(iterations, keylen)
```
Comment (calvinmetcalf) [inline browser.js]: why don't we pull check paramaters into it's own file and require it from both locations

File: browser.js [hunk 7]
```diff
@@ -34,20 +14,33 @@ function checkParameters (iterations, keylen) {
     throw new TypeError('Key length not a number')
   }
 
-  if (keylen < 0 || keylen > MAX_ALLOC) {
+  if (keylen < 0 || keylen > MAX_ALLOC || keylen !== keylen) { /* eslint no-self-compare: 0 */
```
Comment (calvinmetcalf) [inline browser.js]: nice didn't realize that was a standard lint rule

File: browser.js (hunk 6 above)
Comment (fanatid) [inline browser.js]: it's a good idea, I'll do this

File: browser.js (hunk 2 above)
Comment (fanatid) [inline browser.js]: I understand, in this case, do you accept PR to `create-hmac` with `require('create-hash')` instead `require('create-hash/browser')`?

File: node-shim.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    digest = 'sha1'
+  }
+
+  if (digest === 'sha1') return crypto.pbkdf2(password, salt, iterations, keylen, callback)
+
+  if (!Buffer.isBuffer(password)) password = new Buffer(password, 'binary')
+  if (!Buffer.isBuffer(salt)) salt = new Buffer(salt, 'binary')
+
+  checkParameters(iterations, keylen)
+  var child = fork(path.resolve(__dirname, 'node-shim-async.js'))
```
Comment (calvinmetcalf) [inline node-shim.js]: this needs to be changed back

**Comment by calvinmetcalf** [comment]
get rid of the browser._createHmac stuff and fix the typo and I'm good to merge this thank you for the work

**Comment by fanatid** [comment]
@calvinmetcalf done. Thank you that checked this!

**Comment by fanatid** [comment]
Please rename `async-shim.js` to `node-shim-async.js` after merge.
